testing/txsim/account.go

```diff
@@ -0,0 +1,441 @@
+// the highest balance as the master account. Accounts that don't yet exist on chain are
+// ignored.
+func (am *AccountManager) setupMasterAccount(ctx context.Context) error {
+	records, err := am.keys.List()
+	if err != nil {
+		return err
+	}
+
+	for _, record := range records {
+		address, err := record.GetAddress()
+		if err != nil {
+			return fmt.Errorf("error getting address for account %s: %w", record.Name, err)
+		}
+
+		// search for the account on chain
+		balance, err := am.getBalance(ctx, address)
+		if err != nil {
+			log.Err(err).Str("account", record.Name).Msg("error getting initial account balance")
+			continue
+		}
+
+		// the master account is the account with the highest balance
+		if am.masterAccount == nil || balance > am.masterAccount.Balance {
+			accountNumber, sequence, err := am.getAccountDetails(ctx, address)
+			if err != nil {
+				log.Err(err).Str("account", record.Name).Msg("error getting initial account details")
+				continue
+			}
+			pk, err := record.GetPubKey()
+			if err != nil {
+				return fmt.Errorf("error getting public key for account %s: %w", record.Name, err)
+			}
+			am.masterAccount = &Account{
+				Address:       address,
+				PubKey:        pk,
+				Sequence:      sequence,
+				AccountNumber: accountNumber,
+				Balance:       balance,
+			}
+		}
+	}
+
+	if am.masterAccount == nil {
+		return fmt.Errorf("no suitable account found")
+	}
+
+	return nil
+}
+
+// AllocateAccounts is used by sequences to specify the number of accounts
+// and the balance of each of those accounts. Not concurrently safe.
+func (am *AccountManager) AllocateAccounts(n, balance int) []types.AccAddress {
+	if n < 1 {
+		panic("n must be greater than 0")
+	}
+	if balance < 1 {
+		panic("balance must be greater than 0")
+	}
+
+	path := hd.CreateHDPath(118, 0, 0).String()
+	addresses := make([]types.AccAddress, n)
+	for i := 0; i < n; i++ {
+		record, _, err := am.keys.NewMnemonic(am.nextAccountName(), keyring.English, path, keyring.DefaultBIP39Passphrase, hd.Secp256k1)
+		if err != nil {
+			panic(err)
+		}
+		addresses[i], err = record.GetAddress()
+		if err != nil {
+			panic(err)
+		}
+
+		pk, err := record.GetPubKey()
+		if err != nil {
+			panic(err)
+		}
+
+		am.pending = append(am.pending, &Account{
+			Address: addresses[i],
+			PubKey:  pk,
+			Balance: int64(balance),
+		})
+	}
+	return addresses
+}
+
+func (am *AccountManager) Submit(ctx context.Context, op Operation) error {
+	for _, msg := range op.Msgs {
+		if err := msg.ValidateBasic(); err != nil {
+			return fmt.Errorf("error validating message: %w", err)
+		}
+	}
+
+	// create the tx builder and add the messages
+	builder := am.tx.Tx()
+	err := builder.SetMsgs(op.Msgs...)
+	if err != nil {
+		return fmt.Errorf("error setting messages: %w", err)
+	}
+
+	builder.SetFeeAmount(types.NewCoins(types.NewInt64Coin(app.BondDenom, feeAmount)))
+	// the master account is responsible for paying the fees
+	builder.SetFeeGranter(am.masterAccount.Address)
```

Comment:
is there a benefit to using fee granter for every tx instead of funding the accounts in some other mechansim (either by including them in the genesis via the existing tooling in testnode or just sending them funds)?

Reply:
It's a valid point. Perhaps it does make the most sense if we leak that complexity to the sequence developers so that they have to over fund their accounts and manage the gas limits (we can hardcode the gas price to the minimum)